src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -1091,12 +1091,15 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           @inline def tpdPos(transformed: Tree) = typedPos(tree.pos, mode, pt)(transformed)
           @inline def tpd(transformed: Tree)    = typed(transformed, mode, pt)
 
-          @inline def warnValueDiscard(): Unit = if (!isPastTyper && settings.warnValueDiscard) {
+          @inline def warnValueDiscard(): Unit = if (!isPastTyper) {
             def isThisTypeResult = (tree, tree.tpe) match {
               case (Apply(Select(receiver, _), _), SingleType(_, sym)) => sym == receiver.symbol
               case _ => false
             }
-            if (!isThisTypeResult) context.warning(tree.pos, "discarded non-Unit value")
+            if (tree.tpe.termSymbol == UnitModule)
+              context.warning(tree.pos, "discarded companion object of Unit. Write '()' for the unit value, the only value of type Unit.")
```

Comment:
I'm wondering whether we need to detect this special case. Maybe it's enough to improve the warning message? Maybe an addendum like s"To avoid this warning, change the expected type to ${tree.tpe} or explicitly return the Unit value, which is written as `()`.".

(PS: I'd rather not add more lazy vals to Definitions. You could use `tree.symbol == UnitClass.companionModule` to avoid introducing `UnitModule` in definitions, maybe with a name-based guard if we're worried about performance.)

Reply:
I can understand backing out the Definitions addition; and I don't think performance suffers, because no right-minded discards their valuable values.

I would suggest avoiding long and wrappy messages, especially in advance of backported improved messages from dotty. The long and wrappy code that leads to your door, etc.

The improved messaging system could add: `Since you are obviously a rank amateur, or perhaps a hobbyist, you may be unaware that...` etc.